src/github.com/matrix-org/go-neb/services/slackapi/message.go

```diff
@@ -0,0 +1,219 @@
+package slackapi
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"github.com/matrix-org/go-neb/matrix"
+	"github.com/russross/blackfriday"
+	"html/template"
+	"io/ioutil"
+	"mime"
+	"net/http"
+	"regexp"
+	"time"
+)
+
+type slackAttachment struct {
+	Fallback         string `json:"fallback"`
+	FallbackRendered template.HTML
+	Color            *string `json:"color"`
+	ColorRendered    template.HTMLAttr
+	Pretext          string `json:"pretext"`
+	PretextRendered  template.HTML
+
+	AuthorName    *string      `json:"author_name"`
+	AuthorLink    template.URL `json:"author_link"`
+	AuthorIcon    *string      `json:"author_icon"`
+	AuthorIconURL template.URL
+
+	Title     *string `json:"title"`
+	TitleLink *string `json:"title_link"`
+
+	Text         string `json:"text"`
+	TextRendered template.HTML
+
+	MrkdwnIn []string `json:"mrkdwn_in"`
+	Ts       *int64   `json:"ts"`
+}
+
+type slackMessage struct {
+	Text         string `json:"text"`
+	TextRendered template.HTML
+	Username     string            `json:"username"`
+	Channel      string            `json:"channel"`
+	Mrkdwn       *bool             `json:"mrkdwn"`
+	Attachments  []slackAttachment `json:"attachments"`
+}
+
+// We use text.template because any fields of any attachments could
+// be Markdown, so it's convenient to escape on a field-by field basis.
+// We do not do this yet, since it's assumed that clients also escape the content we send them.
+var htmlTemplate, _ = template.New("htmlTemplate").Parse(`
+<strong>@{{ .Username }}</strong> via <strong>#{{ .Channel }}</strong><br />
+{{- with (or .TextRendered .Text nil) }}
+	{{- if . }}
+		{{- . }}<br />
+	{{- end }}
+{{- end }}
+{{- range .Attachments }}
+		{{- if .AuthorName }}
+			{{- if .AuthorLink }}<a href="{{ .AuthorLink }}">{{ end }}
+				{{- if .AuthorIconUrl }}<img src="{{ .AuthorIconUrl }}" />{{ end }}
+				{{- .AuthorName }}
+			{{- if .AuthorLink }}</a>{{ end }}
+			<br />
+		{{- end }}
+	<strong>
+		<font color="{{- .ColorRendered }}">▌</font>
+		{{- if .TitleLink }}
+			<a href="{{ .TitleLink}}">{{ .Title }}</a>
+		{{- else }}
+			{{- .Title }}
+		{{- end }}
+		<br />
+	</strong>
+	{{- if .Pretext }}{{ or .PretextRendered .Pretext }}<br />{{ end }}
+	{{- if .Text }}{{ or .TextRendered .Text }}<br />{{ end }}
+{{- end }}
+`)
+
+var netClient = &http.Client{
+	Timeout: time.Second * 10,
+}
+
+var linkRegex, _ = regexp.Compile("<([^|]+)(\\|([^>]+))?>")
```

Comment:
I don't know about you, but I look at this and I think "what does this do?" - Can you comment what this is doing please? E.g:
```go
// Extract hostname
// https://foobar.com/example/string =>  foobar.com 
var linkRegex, _ = regexp.Compile("<([^|]+)(\\|([^>]+))?>")
```